vsr: standby replicas

matklad opened the pull request:
Add ability to run a tigerbeetle cluster where some of the nodes are
standbys (the configuration is static).

This isn't particularly useful for any practical purposes (well, I'd
imagine you can already do a useful canary that way), but is a
reasonable stepping stone towards reconfig.

CLI wise, this looks like this:

```console
$ ./tigerbeetle format --cluster=0 --replica=0 0_0.tigerbeetle # active
$ ./tigerbeetle format --cluster=0 --replica=1 0_1.tigerbeetle # standby

$ ./tigerbeetle start --addresses=3001,3002 --standby-count=1 0_0.tigerbeetle # active
$ ./tigerbeetle start --addresses=3001,3002 --standby-count=1 0_1.tigerbeetle # standby
```

That is, we pass the number of standbys to start, the first
addresses.len - standby_count replicas are active, the rest are
standbys.

Internally, the numbering scheme is that replicas 0..<replica_count are
active, and replica_count..<replica_count + standby_count are standbys.
Keeping the numbers sequential helps with keeping stuff in arrays. I
expect our ID scheme to change in the future to accommodate dynamic
standbys, so I want to choose the simplest option for now.

Implementation wise, the strategy is for standbys to closely follow
active replicas, and skip sending messages almost at the last possible
moment. So, eg, standbys will still be computing DVC messages, and will
avoid sending them out. The motivation is two fold:

- to keep the amount of if (self.standby()) branches as small as
  possible, to simplify implementation.
- to keep runtime behavior of standbys close to active replicas, too
  make canaries more efficient.

The most interesting bit of actual new code here is fn replicate, where
I implement "interlinked rings" strategy of replication to the standbys.

## Pre-merge checklist

* [X] I am very sure this PR could not affect performance.

sentientwaffle commented on src/tigerbeetle/cli.zig:
```diff
@@ -159,6 +165,9 @@ pub fn parse_args(allocator: std.mem.Allocator) !Command {
             addresses = parse_flag("--addresses", arg);
+        } else if (mem.startsWith(u8, arg, "--standby-count")) {
+            if (command != .start) fatal("--standby-count: supported only by 'start' command", .{});
+            standby_count = parse_flag("--standby_count", arg);
```
```suggestion
            standby_count = parse_flag("--standby-count", arg);
```

matklad replied:
Eagle eye!

matklad commented:
Got it to the point where I can run one active replica and one standby, and standby seems to be aware of what's happening with the cluster! Though I want to push `if (!self.standby())` checks deeper than they currently are. And yeah, `if replica_count == 1` checks already get in a way, because replica_count _is_ one on the standby.

matklad commented:
And now it is actually "works" with the simulator, where works = "run three times OK, and on the fourth time crashed due to pre-existing state-transfer assert, at which point I declared victory". 

Will think about what to land here actually...

matklad commented:
Ok, the plan:

Ok, so I think I basically want to land roughly what's already in this PR (but
productionalized quite a bit obviously). That is:

Add ability to run a tigerbeetle cluster where some of the nodes are standbys
(the configuration is static).

This isn't particularly useful for any practical purposes (well, I'd imagine you
can already do a useful canary that way), but is a reasonable stepping stone
towards reconfig.

CLI wise, this would look like this:

```console
$ ./tigerbeetle format --cluster=0 --replica=0 0_0.tigerbeetle # active
$ ./tigerbeetle format --cluster=0 --replica=1 0_1.tigerbeetle # standby

$ ./tigerbeetle start --addresses=3001,3002 --standby-count=1 0_0.tigerbeetle # active
$ ./tigerbeetle start --addresses=3001,3002 --standby-count=1 0_1.tigerbeetle # standby
```

That is, we pass the number of standbys to `start`, the first `addresses.len -
standby_count` replicas are active, the rest are standbys.

Alternative, we can do something like

```console
$ ./tigerbeetle format --cluster=0 --replica=0 0_0.tigerbeetle
$ ./tigerbeetle format --cluster=0 --standby=1 0_1.tigerbeetle

$ ./tigerbeetle start --addresses=3001 --standby-addresses=3002 0_0.tigerbeetle
$ ./tigerbeetle start --addresses=3001 --standby-addresses=3002 0_1.tigerbeetle
```

This looks a bit more production ready, but is more complex. As I don't think we
ship either CLI as is, it makes sense to go for a simpler option for now.

Internally, the numbering scheme is that replicas `0..<replica_count` are
active, and `replica_count..<replica_count + standby_count` are standbys.
Keeping the numbers sequential helps with keeping stuff in arrays. An
alternative would be to say that standby replicas have their highest bit set.
This feels a bit more reliable (harder to confuse active and standby by
accident), but also more fiddly to program. I expect our ID scheme to change in
the future to accommodate dynamic standbys, so I want to choose a simple option
here.

Implementation wise, the strategy is for standbys to closely follow active
replicas, and skip sending messages almost at the last possible moment. So, eg,
standbys will still be computing DVC messages, and will avoid sending them out.
The motivation is two fold:

- to keep the amount of `if (self.standby())` branches as small as possible, to
  simplify implementation.
- to keep runtime behavior of standbys close to active replicas, too make
  canaries more efficient.

I think for this first request I'll avoid touching the clock, so the cluster
time would be computed _including_ replicas. Will fix in the follow up.

The most interesting bit of actual new code here is `fn replicate`, where I
implement "interlinked rings" strategy of replication to the standbys.


If that's good, I'll work on making this into non-draft!

matklad committed:
vsr: standby replicas

Add ability to run a tigerbeetle cluster where some of the nodes are
standbys (the configuration is static).

This isn't particularly useful for any practical purposes (well, I'd
imagine you can already do a useful canary that way), but is a
reasonable stepping stone towards reconfig.

CLI wise, this looks like this:

```console
$ ./tigerbeetle format --cluster=0 --replica=0 0_0.tigerbeetle # active
$ ./tigerbeetle format --cluster=0 --replica=1 0_1.tigerbeetle # standby

$ ./tigerbeetle start --addresses=3001,3002 --standby-count=1 0_0.tigerbeetle # active
$ ./tigerbeetle start --addresses=3001,3002 --standby-count=1 0_1.tigerbeetle # standby
```

That is, we pass the number of standbys to start, the first
addresses.len - standby_count replicas are active, the rest are
standbys.

Internally, the numbering scheme is that replicas 0..<replica_count are
active, and replica_count..<replica_count + standby_count are standbys.
Keeping the numbers sequential helps with keeping stuff in arrays. I
expect our ID scheme to change in the future to accommodate dynamic
standbys, so I want to choose the simplest option for now.

Implementation wise, the strategy is for standbys to closely follow
active replicas, and skip sending messages almost at the last possible
moment. So, eg, standbys will still be computing DVC messages, and will
avoid sending them out. The motivation is two fold:

- to keep the amount of if (self.standby()) branches as small as
  possible, to simplify implementation.
- to keep runtime behavior of standbys close to active replicas, too
  make canaries more efficient.

The most interesting bit of actual new code here is fn replicate, where
I implement "interlinked rings" strategy of replication to the standbys.

matklad committed:
simulator support for standbys

matklad commented on src/testing/cluster.zig:
```diff
@@ -77,9 +78,12 @@ pub fn ClusterType(comptime StateMachineType: fn (comptime Storage: type, compti
         storages: []Storage,
         storage_fault_atlas: *StorageFaultAtlas,
+        /// NB: includes both active replicas and standbys. 
         replicas: []Replica,
```
Torn a bit about this one. We _can_ rename this and related fields/functions to `nodes_`, but, as the type here is `Replica`, that looks ugly. And I don't want to rename `Replica` to `Node`.

matklad commented on src/vsr/replica.zig:
```diff
@@ -522,9 +538,14 @@ pub fn ReplicaType(
             const replica_count = options.replica_count;
+            const standby_count = options.standby_count;
+            const node_count = replica_count + standby_count;
+            assert(0 < replica_count and replica_count <= constants.replicas_max);
```
I like writing Python-like `lo < x < hi` as `lo < x and x < hi`.

sentientwaffle replied:
With assertions an alternative is to split them up:
```zig
assert(replica_count > 0);
assert(replica_count <= constants.replicas_max);
```
which as a bonus gives more granular panics.

(I'm fine with `lo < x and x < hi` too though).

matklad replied:
Good point re granularity, fixed!

matklad commented on src/vsr/replica.zig:
```diff
@@ -1154,7 +1175,9 @@ pub fn ReplicaType(
                 self.normal_heartbeat_timeout.reset();
-                self.start_view_change_from_all_replicas.unset(self.replica);
+                if (!self.standby()) {
+                    self.start_view_change_from_all_replicas.unset(self.replica);
```
This one is a bit ugly, this is where "not doing something as late as possible breaks": quorum counters a bitsets of `replicas_max` size, so, although they are logically unset for all higher indexes and that would be correct, in code we hit panics if `self.replica` points to a standby. 

Alternatives: 

* guard bitset ops with `if (!self.standby())` (what I ended up doing)
* extend `QuorumCounter`'s size to `nodes_max`, rather than `replicas_max` (don't like this, as we do want to assert that we never try to add standby to the quorum)
* for `QuorumCounter`, wrap the bitset into a type which knows replicas_count and standby_count, which asserts that we never try to set standby, and we never use any inxeses larger than `node_max`. I like this in principle, but I don't like adding extra abstractions.

sentientwaffle replied:
> extend `QuorumCounter`'s size to `nodes_max`, rather than `replicas_max` (don't like this, as we do want to assert that we never try to add standby to the quorum)

This is ugly, but could live in each quorum reset function:

```zig
assert((self.start_view_change_from_all_replicas.iterator(.{ .direction = .reverse }).next() orelse 0) < self.replica_count);
```

(Though it would only discover adding standbys after the fact.)

matklad replied:
:thinking: I am sticking with the current approach --- will re-visit once we have reconfig, that might affect how we think about quorums (though hopefully it _shouldn't_ , as quorums are always relative to a particular epoch)

matklad commented on src/vsr/replica.zig:
```diff
@@ -1449,7 +1472,7 @@ pub fn ReplicaType(
             if (self.ignore_repair_message(message)) return;
 
-            assert(self.replica_count > 1);
+            assert(self.node_count > 1);
```
This is not `!self.sole_replica()`, because now cluster-of-ones actually _can_ get request_prepares from standbys

matklad commented on src/vsr/replica.zig:
```diff
@@ -1918,9 +1941,11 @@ pub fn ReplicaType(
+            assert(!self.sole_replica());
             self.start_view_change_window_timeout.stop();
 
+            if (self.standby()) return;
```
Again, return a bit too early due to bitset

matklad commented on src/vsr/replica.zig:
```diff
@@ -1930,7 +1955,9 @@ pub fn ReplicaType(
-            if (self.replica_count == 1) return;
+
+            if (self.sole_replica()) return;
+            if (self.standby()) return;
```
ditto

matklad commented on src/vsr/replica.zig:
```diff
@@ -4557,16 +4655,55 @@ pub fn ReplicaType(
+                        break :next next_standby;
+                    }
+                }
+            };
```
Open to suggestions how to make this flatter!

sentientwaffle replied:
Untested:

```zig
const next = next: {
    if (!self.standby()) {
        // First, replicate to the next active replica.
        const next_replica = @mod(self.replica + 1, self.replica_count);
        if (next_replica != self.primary_index(message.header.view)) {
            break :next next_replica;
        }
    }

    // Having replicated to all active replicas, stop here or jump to standbys, if any.
    if (self.standby_count > 0) {
        // Select the first standby based on view, just to rotate things sometimes.
        const first_standby = self.standby_index_to_replica(message.header.view);
        if (!self.standby()) break :next first_standby;

        // Now replicate to the next standby.
        const my_index = self.standby_replica_to_index(self.replica);
        const next_standby = self.standby_index_to_replica(my_index + 1);
        if (next_standby != first_standby) break :next next_standby;
    }

    log.debug("{}: replicate: not replicating (completed)", .{self.replica});
    return;
};
```

matklad replied:
Oh, love how you merged the two returns!

matklad commented on src/vsr/replica.zig:
```diff
@@ -4838,11 +4993,14 @@ pub fn ReplicaType(
                 },
                 .prepare => {
+                    maybe(self.standby());
+                    if (replica == self.replica) assert(self.primary());
```
I _think_ this assert would be correct on main as well: only primary sends prepares to itself.

sentientwaffle replied:
Maybe even `assert(self.replica != replica);`. The primary does "send" itself a prepare, but it doesn't go through the "send message" path: https://github.com/tigerbeetledb/tigerbeetle/blob/befe7aa8967247528337d2defd0fd1148b8363cb/src/vsr/replica.zig#L3813

sentientwaffle commented on src/testing/cluster/network.zig:
```diff
@@ -58,11 +58,12 @@ pub const Network = struct {
     pub fn init(
         allocator: std.mem.Allocator,
-        replica_count: u8,
+        node_count: u8,
         client_count: u8,
```
I don't remember why `replica_count`/`client_count` were passed in when we could just pull them from `options`.
May as well do that rather than having to assert that they match anyway.

matklad replied:
Yeah, will do in a follow up --- I feel like I might want to reshuffle casting story here as well, don't want to make current diff bigger.

sentientwaffle commented on src/tigerbeetle/cli.zig:
```diff
@@ -281,6 +308,17 @@ fn parse_addresses(allocator: std.mem.Allocator, raw_addresses: []const u8) []ne
+        error.InvalidCharacter => fatal("--standby-count: value contains an invalid character", .{}),
+    };
+    if (count > constants.standbys_max) {
+        fatal("--standby-count: too many standbys, at most {}ar alowed", .{constants.standbys_max});
```
```suggestion
        fatal("--standby-count: too many standbys; at most {} are allowed", .{constants.standbys_max});
```

sentientwaffle commented on src/tigerbeetle/cli.zig:
```diff
@@ -21,7 +21,7 @@ const usage = fmt.comptimePrint(
     \\  tigerbeetle format --cluster=<integer> --replica=<index> <path>
     \\
-    \\  tigerbeetle start --addresses=<addresses> <path>
+    \\  tigerbeetle start --addresses=<addresses> --standby-count=<integer> <path>
```
In the nebulous world of CLI documentation conventions, doesn't `--standby-count=<integer>` usually indicate a required argument? I think `[--standby-count=<integer>]` would indicate an optional argument.

matklad replied:
I now recall that jorangreef mentioned that we actually deliberately _don't_ have any optional arguments, and require all defaults to be specified explicitly (avoids surprises, and also doesn't _really_ reduce experience, as tigerbeetle is run by some automation, rather than manually).

Something to revisit when we converge on the final way to specify standbys and such.

matklad replied:
Actually, it'll be gone much earlier than that -- when I add replica_cont to format.

sentientwaffle commented on src/tigerbeetle/cli.zig:
```diff
@@ -55,6 +55,11 @@ const usage = fmt.comptimePrint(
+    \\  --standby-count=<integer>
+    \\        Set the number of standby replicas.
+    \\        The last "standby-count" replicas will passively follow the cluster,
+    \\        without participating in consensus.
```
```suggestion
    \\        The last "standby-count" replicas in the "addresses" will passively
    \\        follow the cluster, without participating in consensus.
```

sentientwaffle commented on src/tigerbeetle/cli.zig:
```diff
@@ -55,6 +55,11 @@ const usage = fmt.comptimePrint(
     \\        will be used.
     \\        "addresses[i]" corresponds to replica "i".
     \\
+    \\  --standby-count=<integer>
```
Out of scope for this PR, but:

What happens if some replicas are started believing a particular replica is a standby, while others are started believing it is active?

matklad replied:
Good question: I think the behavior is the same as when you start replicas with different lengths of addresses: replicas disagree about the configuration, and so the behavior is undefined. 

:thinking: I wonder if `epoch` should be a hash of configuration, rather than a counter? That seems like a conceptually nice way to solve this kind of problem. Though, I _think_ we might need ordering on the epochs, so _just_ a hash won't work.

matklad commented on src/vsr/replica.zig:
```diff
@@ -87,7 +88,18 @@ pub fn ReplicaType(
+        /// Total amount of nodes (replicas and standbys) in the cluster.
+        ///
+        /// Invariant: node_count = replica_count + standby_count
+        node_count: u8,
```
An alternative would be to add `fn node_count` instead of materializing the field, because "single source of truth", but the field seems more readable to me (its presence explains replica_count and standby_count)

sentientwaffle replied:
Either works, we have a mix of both already.

matklad commented on src/vsr/replica.zig:
```diff
@@ -4557,16 +4656,56 @@ pub fn ReplicaType(
+                if (!self.standby()) {
+                    // First, replicate to the next active replica.
+                    const next_replica = @mod(self.replica + 1, self.replica_count);
+                    if (next_replica == self.primary_index(message.header.view)) {
```
Note for tomorrow’s self: flip this branch so that the source code order correspond to how the replication actually is going.

matklad commented:
I must say `acceptor + standby = replica` terminology grows on me! Esp because a replica can be *both* an acceptor and a standby in different epochs. I will try this naming scheme, but after this PR lands

sentientwaffle commented on src/vsr/replica.zig:
```diff
@@ -4557,16 +4656,56 @@ pub fn ReplicaType(
+                }
+            };
+            assert(next != self.replica);
+            assert(next < self.node_count);
```
Standby never replicates to acceptor.
```suggestion
            assert(next < self.node_count);
            assert(!(next < self.replica_count and self.standby()));
```

sentientwaffle commented on src/vsr/replica.zig:
```diff
@@ -4557,16 +4656,56 @@ pub fn ReplicaType(
+        fn standby_index_to_replica(self: *const Self, index: u32) u8 {
+            assert(self.standby_count > 0);
+            return self.replica_count + @intCast(u8, @mod(index, self.standby_count));
+        }
```
```suggestion
        }

```

(missing newline)

sentientwaffle commented on src/vsr/replica.zig:
```diff
@@ -4666,6 +4805,8 @@ pub fn ReplicaType(
                     header.checksum,
                 });
 
+                if (self.standby()) return;
```
I was imagining that these "only send if !standby" guards would be in `send_message_to_replica`, but this is even better since we can assert there instead. :+1:

matklad replied:
Yeah, I love how these asserts and ifs line up, and are also cross-checked by "message misdirected" logs. :thinking: might want to check that we don't have misdirected messages in VOPR.

matklad replied:
>Yeah, I love how these asserts and ifs line up, and **are also cross-checked by "message misdirected" logs**.

Lol, I actually missed that!

sentientwaffle commented on src/vsr/replica.zig:
```diff
@@ -4716,6 +4857,8 @@ pub fn ReplicaType(
         fn send_start_view_change(self: *Self) void {
             assert(self.status == .normal or self.status == .view_change);
```
`assert(!self.sole_replica());`

sentientwaffle commented on src/vsr/replica.zig:
```diff
@@ -4754,6 +4897,8 @@ pub fn ReplicaType(
             assert(message.header.commit == self.commit_min);
```
`assert(!self.sole_replica());` at the beginning of this function. (Alas, github)

matklad commented:
Added a bunch of assert about standby messges to the receiving side here:

https://github.com/tigerbeetledb/tigerbeetle/pull/525

(separte PR, because this one is locked due to merge queue)
